### DISCUSSION TITLE
Customizable timezone for the `timestamp` field type

### DISCUSSION THREAD

**Discussion by JMLX42** [keystonejs/keystone#7327] [Feature requests]
Presenting dates in the user's local timezone can be counter-intuitive.

There are already several issues/discussions around counter-intuitive dates/time management:

- https://github.com/keystonejs/keystone/issues/1984
- https://github.com/keystonejs/keystone/discussions/5331
- https://github.com/keystonejs/keystone/issues/6115#issuecomment-1055788552

My understanding is that most of those problems would be solved by being able to customize/enforce the timezone.
But I did not find a discussion about exactly that. So here it is.

I propose to make timezones customizable.

## Admin UI changes

My proposal is to add a timezone selector in the `DatePicker` component.
By default, the selected timezone would be the one the user's is in (no behavior change).

Javascript offers the required APIs:

- To get the timezone: `Intl.DateTimeFormat().resolvedOptions().timeZone` (ex: `"Europe/Paris"`).
- To get the timezone offset: `new Date().getTimezoneOffset()` (ex: `-60`).

So the Admin UI can feature a dropdown with entries formatted as `$timezone ($offset)` (ex: `Europe/Paris (UTC+1)`).

The frontend code can then be adjusted to take the selected value into account before passing it to the backend.

## Database changes

Currently the database stores the timestamps in UTC. Which I think is the way to go.

But storing the timezone actually used to enter the timestamp might be important for consistency. Let's consider the following scenario:

1. The user in timezone UTC+1 selects a date and the UTC-4 timezone then save the changes.
2. The item's page show that date, but since it does not remember which timezone the user actually selected, the whole date/time is different from what the user just entered.

Using the `DatePicker` component for dates might be even more unsettling: since the time is left to 00:00 in this use case, users in the negative offset range of UTC might see a different day/month/year than what was previously entered.

Therefore I propose to save the Admin UI "display" timezone in the database to maintain consistency between edits.

## Field API changes

Since timezones are heavily user-related, I propose to add the following fields to the `timestamp` field type:

- `timezone.default_value`:
  - Either an actual literal [ISO value](https://en.wikipedia.org/wiki/ISO_8601#Time_offsets_from_UTC) (or whaterver NodeJS can work with natively) or an async function to get such value. The second option can be leveraged to get a potential timezone field from whatever user schema is being used. Thus eventually allowing to customize timezones for each user.
  - That value is used by the Admin UI to override the timezone selector default value.
- `timezone.editable`: a boolean stating whether the timezone can actually be set via the Admin UI (`true`), or if `timezone.default_value` is enforced (`false`).

**Comment by JMLX42** [comment]
Now that the date picker issue is resolved, merged and released (cf https://github.com/keystonejs/keystone/releases/tag/2022-03-25), maybe it is a good time to start addressing this?

**Comment by molomby** [comment]
This is a great idea – you often want to store, not an absolute point in time (a `timestamp`), but "wall clock time" in a particular place/time zone. Also discussed in [#2938: Build a "datetime + time zone" field type (`LocalDateTime`?)](https://github.com/keystonejs/keystone/issues/2938), though that was written in the context of Keystone 5 so the type names used are a confusing now (eg. the `DateTimeUtc` type is what became the Keystone 6 `timestamp` type).

Partly, we just haven't gotten to this work yet but another thing holding it back is that it's actually kind of difficult. Time (and time zones in particular) get more complicated the closer you look at them and anything Keystone provides in this space needs to make assumptions about what an application needs to remain practical.

Some of my thoughts...

### Store Time Zone Names

I think it's important we store the time zone _name_ (ie. from the [TZ database](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones)), specifically, the one's corresponding to a physical location (eg. `Australia/Sydney` or `Europe/Paris`). This is opposed to storing _offset_ value, measured in hours and minutes, like `-0900` or `+1100`.

Not only are the named time zones more stable but the mapping to offsets is not one-to-one. You can convert a named time zone to an offset with relative ease but it is a lossy operation – it's usually not possible to convert back the other way (and you need the time zone names for the UI).

### One Field Type or Two?

For a long time there's been an idea to build a dedicated `TimeZone` field (see [#2937: Build a TimeZone field type](https://github.com/keystonejs/keystone/issues/2937)). This would be useful when, for example, recording the time zone that users exist in, or physical places like offices or stores, etc. The current solution of displaying dates in the users current time zone is fine if the user is using the system right now (ie. you can get their time zone from the browser) but what if you want to send them an email?

Assuming we had a stand-alone `TimeZone` field type, is it then also worth the effort to build a field type that combined a "local" time _and_ time zone in a single field, given that you get much of the benefit just using two fields separately? The former maybe lets you produce a nicer interface in the admin UI (and GraphQL for that matter) but the latter is more flexible approach and requires less effort to build.

### Store Local Time or UTC?

I think I make a good argument in #2938 that [the UI should let users enter a local time and a time zone name](https://github.com/keystonejs/keystone/issues/2938#issuecomment-627028393) but what should be _stored_? As I see it (and if you agree we shouldn't store offsets) we have two options:

* Store the time zone name and the point in time (ie. a timestamp in UTC) corresponding to the local time, or.. 
* Store the time zone name and the local time entered

In most cases these approaches are equivalent and behave the same but they diverge anytime the _offset rules for a time zone change_. These don't happen _super_ often but they do happen often enough that we need to account for them.
The US provides are great example of this with its [recent changes to DLS time](https://en.wikipedia.org/wiki/Permanent_time_observation_in_the_United_States#Permanent_daylight_saving_time), signed into law March 15, 2022, that take effect in Nov 2023. 

Let's imagine I was especially organised in Jan this year (2022) and booked a venue for the "2023 US West Coast Keystone Christmas Party". I expect people to come after work so booked it for 5:30 PM, on Friday Dec 15th 2023 at a pizza place in San Fan. I store the start date in a Keystone app by entering the local time `2023-12-15T17:30:00` and the US Pacific time zone, which the TZ database calls `US/Pacific` (or canonically `America/Los_Angeles`). Now, when this data was created, the world was expecting the local time on that date in San Fran to be in Pacific Standard Time (PST), `UTC-0800`, but later, when the laws changed in March 2022, the offset for this date/location shifted to the now permanent Pacific Daylight Time (PDT), `UTC-0700`. So what happens to my stored value?

If we store the datetime in UTC, the conversion from local time to UTC _occurred when the value was entered_, so the `-0800` offset was applied, giving us `2023-12-16T01:30:00z`. Now that the offset rules have changed, when we render the date back into local time we get `2023-12-15T18:30:00` (in PDT `-0700`), or 6:30 PM. This is "technically correct", in that it's represents the same moment in time, but that's not going to help me when I show up hour late for the party and all the pizza is gone.

The solution in this case is to not store UTC at all, but to store the _local time_ (without an offset or time zone info of its own) and the _time zone name_. This lets us apply an offset as needed, if we want to render it in another time zone, _when the value is used_, allowing changes in the local offset to be incorporated ([assuming the node db has been updated](https://stackoverflow.com/questions/55477931/where-does-node-js-get-the-timezone-data-from)).

However, as always with time, there's a complication. What if, along with the San Fran pizza party, I made _another_ booking for the Keystone devs in Sydney?

In early in 2022, when I was doing all this planning, 5:30 PM on Friday Dec 15th 2023 in San Fran is 12:30 PM on Saturday in Sydney – lunch time! Wouldn't it be great to get some people together for pizza there as well and all jump on a video chat? So, back in Jan, I _also_ make a booking for a pizza place in Sydney for the _same moment in time_, `2023-12-16T01:30:00`, which I enter as `2023-12-16T12:30:00` and the tz `Australia/Sydney` (which is `+1100` that time of year).

This is all good and well until the US laws and tz offset for that date change. When that happens, are my two events still in sync? Not if I've stored local times. In this case, the other approach works best – if we store UTC and the time zone name for both events they stay in sync – but of course, that still breaks for the US-only event.

And there's the rub. Either way we store this ("local + tz" or "UTC + tz") Keystone is making assumptions about how these values will be used and sometimes, for some use cases, those assumptions will be wrong.

Personally, I think "local + tz" is the most correct solution. The fact is, most of the time, when people say `X time on Y date in Z place` they're not actually referring to a point in time (a timestamp), they're referring to the local time in that place, using whatever offset _will be in effect_ at that time, even though the exact offset isn't know. Regardless of what we do, decision should be a conscious, well informed and the implications documented.

### Folding?

Maybe we can ignore this for now but, when local time transitions out of DST (and goes "back" an hr) a "fold" is created. Times in this fold are ambiguous – they refer to one of two possible moments in time. For example, the local time `2022-04-03T02:30:00` in `Australia/Sydney` "happened" at both `2022-04-02T15:30:00z` then again at `2022-04-02T16:30:00z`. When converting an ambiguous local time into UTC you need to decide which way to fold it; forward or backward. Ideally the API lets its caller specify.

We'll be trying to keep Keystone in line with [Ecma's proposed Temporal standard](https://tc39.es/proposal-temporal/docs/zoneddatetime.html) but (at a glance) I can't see any word there on how this edge case will be handled.

**Comment by jschuur** [comment]
I ran into an issue today where being able to specify which time zone to interpret the timestamp field value would be helpful. I want the user to know that they are entering timestamps in UTC directly, without needing to convert this in their head.

So I'd mention in the field UI `description` that when they're picking should be UTC and it'll skip converting the selected datetime based on the browser's local time zone.
